File: tests/bench_structures.py
"""Benchmarks for data structures: vocabularies, inverted index (pytest-benchmark)."""

import pytest

from kaos_nlp_core.structures import (
    BloomVocabulary,
    FrequencyVocabulary,
    IndexedVocabulary,
    InvertedIndex,
    SetVocabulary,
)

# --- Vocabulary insert ---


@pytest.mark.benchmark(group="vocabulary_insert")
def test_set_vocab_insert(benchmark, war_and_peace_words):
    def run():
        v = SetVocabulary()
        for w in war_and_peace_words:
            v.insert(w)
        return v.contains("prince")

    benchmark(run)


@pytest.mark.benchmark(group="vocabulary_insert")
def test_freq_vocab_insert(benchmark, war_and_peace_words):
    def run():
        v = FrequencyVocabulary()
        for w in war_and_peace_words:
            v.insert(w)
        return v.contains("prince")

    benchmark(run)


@pytest.mark.benchmark(group="vocabulary_insert")
def test_indexed_vocab_insert(benchmark, war_and_peace_words):
    def run():
        v = IndexedVocabulary()
        for w in war_and_peace_words:
            v.insert(w)
        return v.contains("prince")

    benchmark(run)


@pytest.mark.benchmark(group="vocabulary_insert")
def test_bloom_vocab_insert(benchmark, war_and_peace_words):
    def run():
        v = BloomVocabulary(50000, 0.01)
        for w in war_and_peace_words:
            v.insert(w)
        return v.contains("prince")

    benchmark(run)


# --- Vocabulary lookup ---


LOOKUP_QUERIES = [
    "prince",
    "war",
    "peace",
    "love",
    "death",
    "soldier",
    "napoleon",
    "nonexistent_word_xyz",
    "another_missing_word",
    "zzzzzzz",
]


@pytest.fixture(scope="module")
def built_vocabs(war_and_peace_words):
    """Pre-built vocabularies for lookup benchmarks."""
    set_v = SetVocabulary()
    freq_v = FrequencyVocabulary()
    idx_v = IndexedVocabulary()
    bloom_v = BloomVocabulary(50000, 0.01)
    for w in war_and_peace_words:
        set_v.insert(w)
        freq_v.insert(w)
        idx_v.insert(w)
        bloom_v.insert(w)
    return {"set": set_v, "freq": freq_v, "indexed": idx_v, "bloom": bloom_v}


@pytest.mark.benchmark(group="vocabulary_lookup")
def test_set_vocab_lookup(benchmark, built_vocabs):
    v = built_vocabs["set"]
    benchmark(lambda: [v.contains(q) for q in LOOKUP_QUERIES])


@pytest.mark.benchmark(group="vocabulary_lookup")
def test_freq_vocab_lookup(benchmark, built_vocabs):
    v = built_vocabs["freq"]
    benchmark(lambda: [v.contains(q) for q in LOOKUP_QUERIES])


@pytest.mark.benchmark(group="vocabulary_lookup")
def test_indexed_vocab_lookup(benchmark, built_vocabs):
    v = built_vocabs["indexed"]
    benchmark(lambda: [v.contains(q) for q in LOOKUP_QUERIES])


@pytest.mark.benchmark(group="vocabulary_lookup")
def test_bloom_vocab_lookup(benchmark, built_vocabs):
    v = built_vocabs["bloom"]
    benchmark(lambda: [v.contains(q) for q in LOOKUP_QUERIES])


# --- FrequencyVocabulary top_n ---


@pytest.mark.benchmark(group="vocabulary_top_n")
@pytest.mark.parametrize("n", [10, 100, 1000], ids=lambda n: f"top_{n}")
def test_freq_vocab_top_n(benchmark, built_vocabs, n):
    v = built_vocabs["freq"]
    benchmark(v.top_n, n)


# --- Inverted Index ---


@pytest.fixture(scope="module")
def inverted_index_docs(war_and_peace_text):
    """Split War and Peace into paragraph-sized documents."""
    return [p.split() for p in war_and_peace_text.split("\n\n") if len(p) > 100]


@pytest.fixture(scope="module")
def built_index(inverted_index_docs):
    """Pre-built inverted index for query benchmarks."""
    idx = InvertedIndex()
    for i, doc in enumerate(inverted_index_docs):
        idx.add_document(i, doc)
    return idx


@pytest.mark.benchmark(group="inverted_index")
def test_inverted_index_build(benchmark, inverted_index_docs):
    def run():
        idx = InvertedIndex()
        for i, doc in enumerate(inverted_index_docs):
            idx.add_document(i, doc)
        return idx.term_count()

    benchmark(run)


@pytest.mark.benchmark(group="inverted_index")
def test_inverted_index_query_and_2(benchmark, built_index):
    benchmark(built_index.query_and, ["Prince", "war"])


@pytest.mark.benchmark(group="inverted_index")
def test_inverted_index_query_or_2(benchmark, built_index):
    benchmark(built_index.query_or, ["Prince", "war"])


@pytest.mark.benchmark(group="inverted_index")
def test_inverted_index_query_and_5(benchmark, built_index):
    benchmark(built_index.query_and, ["the", "and", "was", "Prince", "war"])


@pytest.mark.benchmark(group="inverted_index")
def test_inverted_index_tf_idf(benchmark, built_index):
    terms = ["Prince", "war", "peace", "the", "and"]
    benchmark(lambda: [built_index.tf_idf(t, 0) for t in terms])


# --- BM25 ---


@pytest.mark.benchmark(group="bm25")
def test_bm25_score_2_terms(benchmark, built_index):
    benchmark(built_index.score_bm25, ["Prince", "war"], 0)


@pytest.mark.benchmark(group="bm25")
def test_bm25_score_5_terms(benchmark, built_index):
    benchmark(
        built_index.score_bm25,
        ["Prince", "war", "peace", "love", "death"],
        0,
    )


@pytest.mark.benchmark(group="bm25")
def test_bm25_query_top10_2_terms(benchmark, built_index):
    benchmark(built_index.query_bm25, ["Prince", "war"], 10)


@pytest.mark.benchmark(group="bm25")
def test_bm25_query_top10_5_terms(benchmark, built_index):
    benchmark(
        built_index.query_bm25,
        ["Prince", "war", "peace", "love", "death"],
        10,
    )


@pytest.mark.benchmark(group="bm25")
def test_bm25_query_top100_5_terms(benchmark, built_index):
    benchmark(
        built_index.query_bm25,
        ["Prince", "war", "peace", "love", "death"],
        100,
    )


# --- TF-IDF variants ---


@pytest.mark.benchmark(group="tf_idf_variants")
def test_tf_idf_sublinear_smooth(benchmark, built_index):
    benchmark(
        built_index.score_tf_idf,
        ["Prince", "war", "peace"],
        0,
        "sublinear",
        "smooth",
    )


@pytest.mark.benchmark(group="tf_idf_variants")
def test_tf_idf_query_top10(benchmark, built_index):
    benchmark(
        built_index.query_tf_idf,
        ["Prince", "war", "peace"],
        10,
        "sublinear",
        "smooth",
    )

File: benches/bench_structures.rs
//! Benchmarks for data structures: vocabularies, inverted index.

use criterion::{criterion_group, criterion_main, BenchmarkId, Criterion};
use std::hint::black_box;

use _rust::core::structures::{
    inverted_index::{Bm25Params, IdfWeight, InvertedIndex, TfWeight},
    vocabulary::{BloomVocabulary, FrequencyVocabulary, IndexedVocabulary, SetVocabulary},
};

fn load_words() -> Vec<String> {
    let text = std::fs::read_to_string("tests/fixtures/war_and_peace.txt")
        .unwrap_or_else(|_| "The quick brown fox jumps over the lazy dog. ".repeat(10000));
    text.split_whitespace()
        .map(|w| {
            w.trim_matches(|c: char| !c.is_alphanumeric())
                .to_lowercase()
        })
        .filter(|w| !w.is_empty())
        .collect()
}

fn bench_vocabulary_insert(c: &mut Criterion) {
    let words = load_words();
    let mut group = c.benchmark_group("vocabulary/insert");

    group.bench_function("set", |b| {
        b.iter(|| {
            let mut v = SetVocabulary::new();
            for w in &words {
                v.insert(w);
            }
            black_box(v.len())
        })
    });

    group.bench_function("frequency", |b| {
        b.iter(|| {
            let mut v = FrequencyVocabulary::new();
            for w in &words {
                v.insert(w);
            }
            black_box(v.len())
        })
    });

    group.bench_function("indexed", |b| {
        b.iter(|| {
            let mut v = IndexedVocabulary::new();
            for w in &words {
                v.insert(w);
            }
            black_box(v.len())
        })
    });

    group.bench_function("bloom", |b| {
        b.iter(|| {
            let mut v = BloomVocabulary::new(50000, 0.01);
            for w in &words {
                v.insert(w);
            }
            black_box(v.approx_len())
        })
    });

    group.finish();
}

fn bench_vocabulary_lookup(c: &mut Criterion) {
    let words = load_words();

    // Build vocabs
    let mut set_v = SetVocabulary::new();
    let mut freq_v = FrequencyVocabulary::new();
    let mut idx_v = IndexedVocabulary::new();
    let mut bloom_v = BloomVocabulary::new(50000, 0.01);
    for w in &words {
        set_v.insert(w);
        freq_v.insert(w);
        idx_v.insert(w);
        bloom_v.insert(w);
    }

    let queries = [
        "prince",
        "war",
        "peace",
        "love",
        "death",
        "soldier",
        "napoleon",
        "nonexistent_word_xyz",
        "another_missing_word",
        "zzzzzzz",
    ];

    let mut group = c.benchmark_group("vocabulary/lookup");

    group.bench_function("set", |b| {
        b.iter(|| {
            for q in &queries {
                black_box(set_v.contains(q));
            }
        })
    });

    group.bench_function("frequency", |b| {
        b.iter(|| {
            for q in &queries {
                black_box(freq_v.contains(q));
            }
        })
    });

    group.bench_function("indexed", |b| {
        b.iter(|| {
            for q in &queries {
                black_box(idx_v.contains(q));
            }
        })
    });

    group.bench_function("bloom", |b| {
        b.iter(|| {
            for q in &queries {
                black_box(bloom_v.contains(q));
            }
        })
    });

    group.finish();
}

fn bench_frequency_top_n(c: &mut Criterion) {
    let words = load_words();
    let mut freq_v = FrequencyVocabulary::new();
    for w in &words {
        freq_v.insert(w);
    }

    let mut group = c.benchmark_group("vocabulary/top_n");
    for n in [10, 100, 1000] {
        group.bench_with_input(BenchmarkId::from_parameter(n), &n, |b, &n| {
            b.iter(|| black_box(freq_v.top_n(n)))
        });
    }
    group.finish();
}

fn bench_inverted_index(c: &mut Criterion) {
    let text = std::fs::read_to_string("tests/fixtures/war_and_peace.txt")
        .unwrap_or_else(|_| "The quick brown fox jumps over the lazy dog. ".repeat(10000));

    // Split into ~500-word "documents" (paragraphs)
    let docs: Vec<Vec<&str>> = text
        .split("\n\n")
        .filter(|p| p.len() > 100)
        .map(|p| p.split_whitespace().collect())
        .collect();

    let mut group = c.benchmark_group("inverted_index");

    // Build
    group.bench_function("build", |b| {
        b.iter(|| {
            let mut idx = InvertedIndex::new();
            for (i, doc) in docs.iter().enumerate() {
                idx.add_document(i as u32, doc);
            }
            black_box(idx.term_count())
        })
    });

    // Build once for query benchmarks
    let mut idx = InvertedIndex::new();
    for (i, doc) in docs.iter().enumerate() {
        idx.add_document(i as u32, doc);
    }

    group.bench_function("query_and/2_terms", |b| {
        b.iter(|| black_box(idx.query_and(&["Prince", "war"])))
    });

    group.bench_function("query_or/2_terms", |b| {
        b.iter(|| black_box(idx.query_or(&["Prince", "war"])))
    });

    group.bench_function("query_and/5_terms", |b| {
        b.iter(|| black_box(idx.query_and(&["the", "and", "was", "Prince", "war"])))
    });

    group.bench_function("tf_idf", |b| {
        b.iter(|| {
            for term in ["Prince", "war", "peace", "the", "and"] {
                black_box(idx.tf_idf(term, 0));
            }
        })
    });

    group.finish();
}

fn bench_bm25(c: &mut Criterion) {
    let text = std::fs::read_to_string("tests/fixtures/war_and_peace.txt")
        .unwrap_or_else(|_| "The quick brown fox jumps over the lazy dog. ".repeat(10000));

    let docs: Vec<Vec<&str>> = text
        .split("\n\n")
        .filter(|p| p.len() > 100)
        .map(|p| p.split_whitespace().collect())
        .collect();

    let mut idx = InvertedIndex::new();
    for (i, doc) in docs.iter().enumerate() {
        idx.add_document(i as u32, doc);
    }

    let params = Bm25Params::default();
    let mut group = c.benchmark_group("bm25");

    group.bench_function("score/2_terms", |b| {
        b.iter(|| black_box(idx.score_bm25(&["Prince", "war"], 0, &params)))
    });

    group.bench_function("score/5_terms", |b| {
        b.iter(|| {
            black_box(idx.score_bm25(&["Prince", "war", "peace", "love", "death"], 0, &params))
        })
    });

    group.bench_function("query_top10/2_terms", |b| {
        b.iter(|| black_box(idx.query_bm25(&["Prince", "war"], &params, 10)))
    });

    group.bench_function("query_top10/5_terms", |b| {
        b.iter(|| {
            black_box(idx.query_bm25(&["Prince", "war", "peace", "love", "death"], &params, 10))
        })
    });

    group.bench_function("query_top100/5_terms", |b| {
        b.iter(|| {
            black_box(idx.query_bm25(&["Prince", "war", "peace", "love", "death"], &params, 100))
        })
    });

    group.finish();
}

fn bench_tf_idf_variants(c: &mut Criterion) {
    let text = std::fs::read_to_string("tests/fixtures/war_and_peace.txt")
        .unwrap_or_else(|_| "The quick brown fox jumps over the lazy dog. ".repeat(10000));

    let docs: Vec<Vec<&str>> = text
        .split("\n\n")
        .filter(|p| p.len() > 100)
        .map(|p| p.split_whitespace().collect())
        .collect();

    let mut idx = InvertedIndex::new();
    for (i, doc) in docs.iter().enumerate() {
        idx.add_document(i as u32, doc);
    }

    let mut group = c.benchmark_group("tf_idf_variants");

    group.bench_function("raw_standard", |b| {
        b.iter(|| {
            black_box(idx.score_tf_idf(
                &["Prince", "war", "peace"],
                0,
                TfWeight::Raw,
                IdfWeight::Standard,
            ))
        })
    });

    group.bench_function("sublinear_smooth", |b| {
        b.iter(|| {
            black_box(idx.score_tf_idf(
                &["Prince", "war", "peace"],
                0,
                TfWeight::Sublinear,
                IdfWeight::Smooth,
            ))
        })
    });

    group.bench_function("query_top10/sublinear_smooth", |b| {
        b.iter(|| {
            black_box(idx.query_tf_idf(
                &["Prince", "war", "peace"],
                TfWeight::Sublinear,
                IdfWeight::Smooth,
                10,
            ))
        })
    });

    group.finish();
}

criterion_group!(
    benches,
    bench_vocabulary_insert,
    bench_vocabulary_lookup,
    bench_frequency_top_n,
    bench_inverted_index,
    bench_bm25,
    bench_tf_idf_variants,
);
criterion_main!(benches);
